### PR TITLE
Allow all configuration options of livereload to be passed through.

### LINKED ISSUE #100: Allow options to be passed through to connect-livereload.
Is there any way to specify options for connect-livereload via the grunt-contrib-connect configuration? I'd like to change the livereload ignore list but I can't see a way to achieve it without overriding the middleware.

My apologies if there is something obvious I've missed.

Thanks!

### PR DESCRIPTION
**Pull request by aknrdureegaesr**
Allow all [livereload]() configuration options to be passed through.

Fixes #100

### CODE REVIEW

**Comment by aknrdureegaesr** [comment]
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA256

JS Foundation Individual Contributor License Agreement v4.1
Thank you for your interest in the JS Foundation. This document clarifies the terms under which you, the person listed below, may make Contributions — which may include without limitation, software, bug fixes, configuration changes, documentation, or any other materials — to any of the projects owned or managed by the JS Foundation.

Please complete the following information about you and the Contributions. If you have questions about these terms, please contact us at legal@js.foundation.

You accept and agree to the following terms and conditions for your present and future Contributions submitted to the JS Foundation. Except for the license granted herein to the JS Foundation, you reserve all right, title, and interest in and to your Contributions.

Licenses
JS Foundation projects, including code contributions and all related materials, are each released under the terms of the individual licenses as noted in the project’s repository, or, if no separate license is specified, under the terms of the Apache v2.0 license (the “Applicable License”), with the exception of project documentation and sample code, in accordance with the JS Foundation IP Policy.

All project documentation, including without limitation code that is intended as sample code, will be contributed to the Project and made available under:

the Applicable License; or
if approved by the project and defined in the project’s LICENSE file, the Creative Commons Attribution 4.0 International (CC BY 4.0) License.
If approved by the project and defined in the project’s LICENSE file, sample code may be contributed and licensed separately under the Creative Commons CC0 1.0 Universal (CC0) License, by which all copyright and related rights are waived.

You certify that:
Your Contributions are either:
Created in whole or in part by you and you have the right to submit it under the designated license; or
Based upon previous work that, to the best of your knowledge, is covered under an appropriate open source license and you have the right under that license to submit that work with modifications, whether created in whole or in part by you, under the designated license; or
Provided directly to you by some other person who certified (1) or (2) and you have not modified them.
You understand and agree that the JS Foundation projects and your Contributions are public and that a record of the Contributions (including all metadata and personal information you submit with them) is maintained indefinitely and may be redistributed consistent with the JS Foundation’s policies and the requirements of the Apache v2.0 license where they are relevant.
You are granting your Contributions to the JS Foundation under the terms of the license as noted in the project’s repository or if not specified, either the Apache v2.0 license or the CC0 as applicable.

I, Andreas Krüger, sign this license.  I have enclosed my signature to this end.  My contributions (git-commits) are signed with the same key.

Andreas Krüger, andreas.krueger@innoq.com, Berlin, Germany, 2018-08-23.
-----BEGIN PGP SIGNATURE-----

iQEzBAEBCAAdFiEE0I9nurghE7C7Y8QpsAyWGpDzS7gFAlt+jFIACgkQsAyWGpDz
S7gIYgf/VHixREVeoDjtidTUIjsThGfpVFy0+oTqMM6MIKhxAi9W+AIdYcGKiKYE
G8aiMldE1sEFe8URGx0jDYqAs6Dz6hZyBmcXFJ7k6FASf1tajxoJ7xghHeuVR9p/
LdkKDJeGwcadg3S7isHFO+uRpuLSQmALqRI9T7CrequrlmxHbt0wOhTmpAGvI5LO
NVLVDaOVP9XJvcgBoH8pEm94iE9hrK+Cn4md6PiMSHxqEgptzKWYQOY8/sitSXlU
z1CVoYSzgpE+HR6/qeK2BSTW0RflLQX8G6OdurjJgYUSi6iRu7NoqbcFyE89NAX0
dAjgKkw8q5SmZhuTctH1/vXlcB6Kbg==
=Mmqj
-----END PGP SIGNATURE-----

File: CHANGELOG [hunk 1]
```diff
@@ -1,3 +1,7 @@
+v1.1.0:
```
Comment (XhmikosR) [inline CHANGELOG]: Please leave all version changes to the person who will make the release if this is merged.

**Comment by aknrdureegaesr** [comment]
I have looking for the test to augment by temporarily introducing a bug in the existing passing of the parameters:

```
        // Inject live reload snippet
        if (options.livereload !== false) {
          middleware.unshift(injectLiveReload({port: 35729, hostname: 'localhost'}));
          callback(null);
        } else {
          callback(null);
        }
```

When I run `npm test`, all tests still pass.

Is there any other way to run tests, or does a test that tests parameter passing simply not exist presently?

**Comment by XhmikosR** [comment]
The tests are here https://github.com/gruntjs/grunt-contrib-connect/blob/9da2f3788a1b61b368562e6c879d1dfd292f0057/test/connect_test.js#L193-L218

**Comment by aknrdureegaesr** [comment]
There are three code branches:

* `options.livereload` is `true`
* `options.livereload` is a number
* `options.livereload` is a map.

The redundant `hostname: options.hostname` is only set in two of the three.

**Comment by XhmikosR** [comment]
You could pass your options there and check if the port/hostname etc match.

>The redundant hostname: options.hostname is only set in two of the three.

Yeah, but still, it's duplication.

**Comment by aknrdureegaesr** [comment]
Yes, I've seen those tests. Those tests do not test whether the number passed as a port is actually used, so they test something, but not parameter handover.

I take your answer as "tests that test parameter passing do not presently exist".

I'll see what I can do.

**Comment by aknrdureegaesr** [comment]
The way to avoid it I can think of is something like

```
        if (options.livereload !== false) {
          if (options.livereload === true || typeof options.livereload === 'number') {
            ...
            if (options.livereload === true) {
              ...
            } else {
              ...
            }
          }
```

Seems to me that cure is worse than the disease. What do you think?

**Comment by XhmikosR** [comment]
OK, let's forget about the deduplication, but please try to add a test.

**Comment by aknrdureegaesr** [comment]
> OK, let's forget about the deduplication, but please try to add a test.

Done.

File: test/connect_test.js [hunk 2]
```diff
@@ -242,6 +242,32 @@ exports.connect = {
...
+    test.expect(2);
+    get({
+      hostname: 'localhost',
+      port: 8019,
+      path: '/livereload.html',
+      headers: {
+        accept: 'text/html'
+      }
+    }, function(res, body) {
+      test.ok(body.indexOf('https://example.org:54321/livereload.js?snipver=1') !== -1, 'Should contain livereload snippet with configured src.');
```
Comment (XhmikosR) [inline test/connect_test.js]: Maybe just use `body.includes`?

File: docs/connect-options.md [hunk 3]
```diff
@@ -56,12 +56,14 @@ Default: `false`
...
 ## livereload
-Type: `Boolean` or `Number`  
+Type: `Boolean`, `Number`, or `Object`  
 Default: `false`
 
-Set to `true` or a port number to inject a live reload script tag into your page using [connect-livereload](https://github.com/intesso/connect-livereload).
+Set to anything but `false` to inject a live reload script tag into your page using [connect-livereload](https://github.com/intesso/connect-livereload).
 
-*This does not perform live reloading. It is intended to be used in tandem with grunt-contrib-watch or another task that will trigger a live reload server upon files changing.*
+If you set to `true`, defaults are used. If you set to a number, that number is used as a port number, together with the hostname you configured. If you set this to a map, that map is passed to `connect-livereload` unchanged as its configuration.
```
Comment (XhmikosR) [inline docs/connect-options.md]: map -> object

**Comment by XhmikosR** [comment]
So, @aknrdureegaesr, thanks for the patience so far!

Now, I was wondering, the 2 tests you added do seem indentical in the objects you are passing. Do we need them to be like that or should we test for different things?

**Comment by aknrdureegaesr** [comment]
At this point, we have three cases: Livereload is simply true, or is a number, or is an object. The possibility to add an object is new. The other two configuration options existed before.

As I found [out](https://github.com/gruntjs/grunt-contrib-connect/pull/246#issuecomment-419701005), there was no previous test for the "number" case. There was only a test on the "true" case.

So I [added](https://github.com/gruntjs/grunt-contrib-connect/pull/246/commits/d8c4d3f4eb37309543f04d76e8b99d7c8633969b) that missing test for the previously-existing "number" functionality. And I also added a test for my own work. So we now have three tests instead of one.
The two new tests very closely model what had been done in the previously existing test.

But...

I am not so sure the previously existing functionality is actually useable. At least on my machine, in a real-world example, both variants (true and number) kept producing a script URI along the lines of `http://0.0.0.0:port/...`. My browser could not successfully fetch a script from `0.0.0.0`. The patch I present here is helpful, as it allows to precisely determine the URI.

It might be worth a consideration to simply remove the old "true" and the old "port" configuration. There is nothing you could do with those that you couldn't do with the full way, that my patch introduces.

But it might break compatibility for people who have been using the old way and for whom that worked. I'm not sure whether such people exist or don't. Those people would have to re-configure.

In short: To reduce code and tests, we could remove the two old ways, "true" and number, and only keep my new way. I'd be happy to reduce the code in that way.

What do you think, is better? Small, or compatible?

**Comment by XhmikosR** [comment]
Let's keep the changes to the minimum here and you can always make a new PR to improve tests later.

File: test/connect_test.js (hunk 2 above)
Comment (aknrdureegaesr) [reply test/connect_test.js]: Just for curiosity: You wanted this change - why are you reverting it back now?

Comment (XhmikosR) [reply test/connect_test.js]: Because I thought wrongfully we had switched to ES6 :P
